lua/jdtls/dap.lua

```diff
@@ -372,49 +372,24 @@ local hotcodereplace_type = {
 }
 
 
-local original_configurations = nil
-function M.setup_dap(opts)
-  local status, dap = pcall(require, 'dap')
-  if not status then
-    print('nvim-dap is not available')
-    return
-  end
-  if dap.adapters.java and original_configurations then
-    return
-  end
-  opts = opts or {}
-  dap.listeners.before['event_hotcodereplace']['jdtls'] = function(session, body)
-    if body.changeType == hotcodereplace_type.BUILD_COMPLETE then
-      if opts.hotcodereplace == 'auto' then
-        vim.notify('Applying code changes')
-        session:request('redefineClasses', nil, function(err)
-          assert(not err, vim.inspect(err))
-        end)
-      end
-    elseif body.message then
-      vim.notify(body.message)
-    end
-  end
-
-  dap.adapters.java = start_debug_adapter
-  original_configurations = dap.configurations.java or {}
-  local configurations = vim.deepcopy(original_configurations)
-  dap.configurations.java = configurations
+function M.fetch_main_configs(callback)
+  local configurations = {}
+  util.execute_command({command = 'vscode.java.resolveMainClass'}, function(err, mainclasses)
+    assert(not err, vim.inspect(err))
 
-  util.execute_command({command = 'vscode.java.resolveMainClass'}, function(err0, mainclasses)
-    if err0 then
-      print('Could not resolve mainclasses: ' .. err0.message)
+    local remaining = #mainclasses
+    if remaining == 0 then
+      callback(configurations)
       return
     end
-
     for _, mc in pairs(mainclasses) do
       local mainclass = mc.mainClass
       local project = mc.projectName
-
       with_java_executable(mainclass, project, function(java_exec)
-        util.execute_command({command = 'vscode.java.resolveClasspath', arguments = { mainclass, project }}, function(err2, paths)
-          if err2 then
-            print(string.format('Could not resolve classpath and modulepath for %s/%s: %s', project, mainclass, err2.message))
+        util.execute_command({command = 'vscode.java.resolveClasspath', arguments = { mainclass, project }}, function(err1, paths)
+          remaining = remaining - 1
+          if err1 then
+            print(string.format('Could not resolve classpath and modulepath for %s/%s: %s', project, mainclass, err1.message))
```

Comment:
This would be a change to the business logic, but why not call `callback` on all successfully loaded configurations?
You could log a warning for each configuration for which the `classpath`/`modulepath` could not be resolved.
Or you could add an optional `unresolvable_configurations` argument to the `callback` so that users could perform manual corrections.

Reply:
You mean it should call the callback once for each configuration? 

Currently it will print the message for all classpath/ modulepath entries that couldn't be resolved and call the callback with the configurations it managed to resolve.

The early return is inside the inner callback function.  It doesn't abort the outer loop

Reply:
Ah, I see that now.

Yes, a callback `on_configiration_detected` that is applied to each configuration would reduce the complexity of this function (as it eliminates the need for the `remaining` counter) and of client calls.
An optional `on_resolve_failed` callback could provide more fine-grained control over error handling.

Reply:
Good point. Although it has the disadvantage that you won't know how many times the callbacks will be called,  so you can't be sure when the operation is complete.

Not sure which option is better

Reply:
`execute_command` and `with_java_executable` are both non-blocking/asynchronous. So `M.fetch_main_configs` will return almost immediately and the callback will be triggered later.


I'll go ahead and merge this as is for now - it's always possible to tweak it later if necessary.

Reply:
Ah, okay then that could make it more complicated.
Anyways, thanks for the fix!